metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java

```diff
@@ -0,0 +1,504 @@
+     * version change which needs to be handled differently.
+     *
+     * @param delta     The metadata delta we are preparing.
+     * @param reader    The reader which yields the batches.
+     * @return          A manifest of what was loaded.
+     */
+    LogDeltaManifest loadLogDelta(
+        MetadataDelta delta,
+        BatchReader<ApiMessageAndVersion> reader
+    ) {
+        long startNs = time.nanoseconds();
+        int numBatches = 0;
+        long numBytes = 0L;
+        long lastOffset = image.provenance().offset();
+        int lastEpoch = image.provenance().epoch();
+        long lastContainedLogTimeMs = image.provenance().lastContainedLogTimeMs();
+
+        while (reader.hasNext()) {
+            Batch<ApiMessageAndVersion> batch = reader.next();
+            int indexWithinBatch = 0;
+            for (ApiMessageAndVersion record : batch.records()) {
+                try {
+                    delta.replay(record.message());
+                } catch (MetadataVersionChangeException e) {
+                    handleMetadataVersionChange(delta,
+                        e.change(),
+                        batch.baseOffset() + indexWithinBatch,
+                        lastEpoch,
+                        lastContainedLogTimeMs);
+                } catch (Throwable e) {
+                    faultHandler.handleFault("Error loading metadata log record from offset " +
+                            batch.baseOffset() + indexWithinBatch, e);
+                }
+                lastEpoch = batch.epoch();
+                lastContainedLogTimeMs = batch.appendTimestamp();
+                indexWithinBatch++;
+            }
+            metrics.updateBatchSize(batch.records().size());
+            lastOffset = batch.lastOffset();
+            numBytes += batch.sizeInBytes();
+            numBatches++;
+        }
+        MetadataProvenance provenance =
+                new MetadataProvenance(lastOffset, lastEpoch, lastContainedLogTimeMs);
+        long elapsedNs = time.nanoseconds() - startNs;
+        // TODO: this metric should be renamed something like "delta processing time"
+        metrics.updateBatchProcessingTime(elapsedNs);
+        return new LogDeltaManifest(provenance,
+                numBatches,
+                elapsedNs,
+                numBytes);
+    }
+
+    /**
+     * Handle a change in the metadata version.
+     *
+     * @param delta             The metadata delta we're working with.
+     * @param change            The change we're handling.
+     * @param changeOffset      The offset of the change.
+     * @param preChangeEpoch    The log epoch BEFORE the change.
+     */
+    void handleMetadataVersionChange(
+        MetadataDelta delta,
+        MetadataVersionChange change,
+        long changeOffset,
+        int preChangeEpoch,
+        long lastContainedLogTimeMs
+    ) {
+        // First, we materialize the current metadata image and send it to all the publishers that
```

Comment:
It would be helpful to add some motivation for this. Is it for recovery if something goes wrong after upgrade/downgrade?

Reply:
The issue is that different publishers may be installed at different times.

For example, let's say you have some hypothetical cluster balancing service that can be dynamically enabled. (not included in this PR, of course). When it's dynamically enabled, it registers as a metadata publisher, to get all the updates.

If you enable this thing well after startup, you will need to send it a MetadataDelta that contains all the updates (since all the updates are new to it, although not to the existing publishers). How do you get such a MetadataDelta? Write out the existing image to a delta.

Basically this PR adds dynamic installation / removal to the system, which it didn't previously have. That's why the old code locks it to a single publisher. But that isn't very dynamic and we have some use cases for installing things at runtime.